### PR TITLE
Compute maps up front in SnapshotRetentionTask

### PR DESCRIPTION
**Pull request by DaveCTurner**
Rather than computing a map of all relevant snapshots (per repo and
policy) repeatedly, once for every snapshot, let's compute them all up
front.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/part-2

File: x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java [hunk 1]
```diff
@@ -187,7 +187,7 @@ static Map<String, SnapshotLifecyclePolicy> getAllPoliciesWithRetentionEnabled(f
 
     static boolean snapshotEligibleForDeletion(
         SnapshotInfo snapshot,
-        Map<String, List<SnapshotInfo>> allSnapshots,
+        Map<String, Map<String, Map<SnapshotId, RepositoryData.SnapshotDetails>>> allSnapshotDetails,
```
Comment (joegallo) [inline x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java:190]: You're straining the very definition of data structures here, David. 😆

Comment (DaveCTurner) [reply x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java:190]: If this was Haskell they'd be type synonyms.

**Review by joegallo — APPROVED** [review]
+1, I'm a bit +-0 on the (internal only) mutability of the map of maps of maps, but I'm willing to go along for the ride on this and some follow up PRs and revisit this when the dust settles.

**Pull request merged**